include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -37,15 +37,37 @@ class CSummedAreaTableImageFilterBase
 					const auto regions = inputImage->getRegions();
 
 					size_t retval = {};
-					for (const auto* region = regions.begin(); 
-						region < regions.end(); 
-						++region)
-						retval += region->imageExtent.width * region->imageExtent.height * region->imageExtent.depth * inputImage->getCreationParameters().arrayLayers;
+					for (const auto* region = regions.begin(); region != regions.end(); ++region)
+						retval += (region->bufferRowLength ? region->bufferRowLength : region->imageExtent.width) * region->imageExtent.height * region->imageExtent.depth;
 
-					retval *= channels * decodeByteSize;
+					retval *= inputImage->getCreationParameters().arrayLayers * channels * decodeByteSize;
 
 					return retval;
 				}
+
+				inline auto& getImageTotalCacheBuffer()
+				{
+					return imageTotalCacheBuffersOutput;
+				}
+
+			protected:
+
+				inline void createImageTotalBuffers(const size_t scratchBufferByteSize, const size_t outputBufferByteSize, const uint32_t imageDepth)
+				{
+					imageTotalCacheBuffersScratch.clear();
+					imageTotalCacheBuffersOutput.clear();
+					for (uint32_t i = 0; i < imageDepth; ++i)
+					{
+						imageTotalCacheBuffersScratch.emplace_back(core::make_smart_refctd_ptr<asset::ICPUBuffer>(scratchBufferByteSize));
+						memset(imageTotalCacheBuffersScratch[i]->getPointer(), 0, imageTotalCacheBuffersScratch[i]->getSize());
+
+						imageTotalCacheBuffersOutput.emplace_back(core::make_smart_refctd_ptr<asset::ICPUBuffer>(outputBufferByteSize));
+						memset(imageTotalCacheBuffersOutput[i]->getPointer(), 0, imageTotalCacheBuffersOutput[i]->getSize());
+					}
+				}
+
+				core::vector<core::smart_refctd_ptr<asset::ICPUBuffer>> imageTotalCacheBuffersScratch;			//!< Buffers holding total image SAT values in decoding mode.
+				core::vector<core::smart_refctd_ptr<asset::ICPUBuffer>> imageTotalCacheBuffersOutput;			//!< Buffers holding total image SAT values after encoding to image. A user is responsible to reinterprate the memory correctly.
```

Comment:
no, you're not allowed to allocate extra data... you partition the reserved space and use that.

Reply:
I will use scratch then

Reply:
Well, actually I use the extra cache for columns, but I need to know the width of an image that depends on the chosen mip map level. I can bypass it by allocating just the max possible width in the scratch, but I don't know if we want such an approach. What should I do?

Reply:
allocated in scratch

Reply:
solved